### PR TITLE
multi+refactor: persistent peer manager

### LINKED ISSUE #5377: Peers are unable to reconnect and stuck in a deadlock after IP address change
As described by @t-bast in https://gitter.im/ACINQ/eclair LND and eclair (possibly c-lightning) peers are unable to reconnect after a IP address change due to not trusting `node_announcement`:

> if you change your IP address, your peers won't reconnect to you. They'll wait for you to connect to them and then will store your new IP address for future reconnections (ie they don't trust the new IP address from `node_announcement` alone)

This leads to a situation where even some clearnet peers have to be manually reconnected which can be quite time consuming and TOR peers are stuck in a deadlock that can only manually be resolved by the node operator of each and every TOR peer. They would have to look up the new IP address of their offline peer on one of the lightning explorer websites and manually feed it to their node. This is unlikely to happen quickly, if at all.

I changed the IP of my node more than 2 days ago and still have 65 disconnected TOR peers. Most, if not all of them seem to be running LND.

I'd suggest to use the IP from `node_announcement` if a connection to the last known IP fails. Can still be alternated with last known IP until success.

Also reported to eclair: https://github.com/ACINQ/eclair/issues/1842

### PR DESCRIPTION
**Pull request by ellemouton**
This PR adds a `PersistentPeerManager` to the server and refactors all persistent peer logic to be handled by the `PersistentPeerManager`. The end result is that  `persistentPeers`, `persistentPeersBackoff`, `persistentConnReqs` and `persistentRetryCancels` are all removed from the `server` struct and replaced by a `PersistentPeerManager`.

### CODE REVIEW

File: peer/connmgr.go
Comment (ellemouton) [inline peer/connmgr.go:1]: Note: I will still add unit tests for this. Just waiting for a go-ahead on the approach first

File: lntest/itest/lnd_network_test.go
Comment (alexbosworth) [inline lntest/itest/lnd_network_test.go:64]:
```suggestion
// its listening address then its peer will still be able to reconnect to it.
```
```suggestion
// its listening address then it's peer will still be able to reconnect to it.
```

File: lntest/itest/lnd_network_test.go
Comment (yyforyongyu) [inline lntest/itest/lnd_network_test.go:224]: how about making this into a helper closure?

**Review by yyforyongyu — COMMENTED** [review]
cACK, I like the direction where we start moving code out of the `server.go`, def involves more work tho. My question is, since we are already here, can we move all connection management into peer?

**Comment by ellemouton** [comment]
>  My question is, since we are already here, can we move all connection management into peer?

Very happy to do so! Do you think it should all be in 1 PR?

**Comment by yyforyongyu** [comment]
> Mind elaborating? Do you mean in general or specifically in this PR?

Meant in this PR. Since we are moving the code outside of `server.go`, it's finally a good chance to patch the missing unit tests and refactor some of them if needed. But yeah, it would involve more work.

> Very happy to do so! Do you think it should all be in 1 PR?

It depends. I'd think about how emergent issue #5377 is. If it's a pressing issue, I'd suggest we move to #5538 and land it first, since it's almost done and has been reviewed. If not, we might as well take the chance to refactor the connection manager here, moving it into its own package. This will def take more time as the scale is large, but we should do it imo. Another option is to continue what you have here, that we move out the persistent manager as the first step. Guess it's a judgment call from @Roasbeef .

**Comment by viaj3ro** [comment]
> If it's a pressing issue, I'd suggest we move to #5538 and land it first, since it's almost done and has been reviewed

I'd argue it is pressing. I still have a handful of TOR peers that refuse to connect back to me even after almost 4 months. Since many nodes have IP address changes once in a while, I assume they have the same issue without ever noticing it.

**Comment by ellemouton** [comment]
holding off on this PR until 0.15. Working on the solution in #5538 again instead 👍

**Comment by hsjoberg** [comment]
I'd also argue it's quite pressing, I have several users of my wallet which have encountered this issue.

**Comment by ellemouton** [comment]
!lightninglabs-deploy mute 2022-Jan-01

**Comment by yyforyongyu** [comment]
Is this ready for another round?

**Comment by ellemouton** [comment]
@yyforyongyu, yes it is ready :)

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Good starting point! I'm thinking maybe we should add a dedicated package/service to act as our conn manager, but that's for the long run. I'm halfway through, submitting some questions first.

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:61]: hmm but it does only guard the conns map at the moment..? happy to rename it to mu. but not sure about embedding it

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:75]: no it is the address string. will update comment to make it more clear 👍

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:92]: it is the only way i could get this initialisation logic to work

```
// Initialize a retry canceller for this peer if one does not exist.
	var cancelChan chan struct{}
	if peer.retryCanceller != nil {
		cancelChan = *peer.retryCanceller
	} else {
		cancelChan = make(chan struct{})
		peer.retryCanceller = &cancelChan
	}
```

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:134]: because the subscription must not be canceled when `Start` exits, it must be canceled when the goroutine exits. Start will exit as soon as it has spun off the goroutine

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:146]: I have added a log for now. 
not sure what the best way is to return errors from a goroutine? should the caller of `Start` be given an error channel to watch?

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go]: hehe good idea :)

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go]: oh! thanks. updated 👍

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:236]: we might have added a persistent peer using `AddPeer` with a nil addr map. This is especially needed for the transition between the current state of the peer logic and this PR. But maybe yes, after the transition, we will expect addr to always be non-empty

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:248]: not sure what you mean? this is a temporary map that we can freely delete things from in this function.

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:266]: I think we only ever want 1 connection. see https://github.com/lightning/bolts/blob/master/01-messaging.md#connection-handling-and-multiplexing

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:306]: yes i think so. This is just a duplicate of the `connectToPersistentPeer` logic in `server.go`. That code was recently added but you can see in this diff [here](4d73f966aa68720542932437d688f56b370) that it was also set to `true` before

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:310]: yeah defs. but just added it since this is a goroutine so there is still a change we remove the peer from the map before getting here

**Review by ellemouton — COMMENTED** [review]
Thanks for the initial review @yyforyongyu 🚀 
I have left some comments & addressed some of you questions and did some updates

**Review by joostjager — COMMENTED** [review]
Not totally familiar with this part of the code, but left a few comments. For me, the various goroutines and locks make it difficult to ascertain that the logic is correct. I wonder if it would be easier to understand with a single event loop per peer that receives updates via a channel.

**Comment by ellemouton** [comment]
Hi @joostjager ! Thanks so much for the review on this! Im gonna mute the bot on this PR for a bit just while 0.15 is top prio & will address your review as soon as i can 👍

**Comment by ellemouton** [comment]
!lightninglabs-deploy mute 2022-April-18

File: server.go
Comment (ellemouton) [reply server.go:2821]: will add 👍

File: server.go
Comment (ellemouton) [reply server.go:1175]: yeah good idea

**Comment by ellemouton** [comment]
I will push an update addressing all the comments after the review club tomorrow (I decided to change the structure quite a bit so I don't want to push it before the review club in case people are already familiar with the old structure)

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:431]: cool, added this in a later commit so that this one can remain a refactor

File: peer/persistentpeer_mgr.go
Comment (ellemouton) [reply peer/persistentpeer_mgr.go:602]: I think this should be fine now that any previous retry canceller channel will get closed

File: peer/persistentpeer_mgr.go
Comment (Crypt-iQ) [reply peer/persistentpeer_mgr.go:98]: I think you could use a regular channel and a boolean that, when set, means that the channel was closed. They both achieve the same thing, but I think the bool approach is nicer

**Review by Crypt-iQ — COMMENTED** [review]
checkpointing review here

File: peer/persistentpeer_mgr.go
Comment (joostjager) [inline peer/persistentpeer_mgr.go:501]: I think at this point, there may still be a `ConnectPeer` goroutine blocked on obtaining the lock. After this function returns and unblocks, `ConnectPeer` may still add something to the `connReqs`?

File: peer/persistentpeer_mgr.go
Comment (joostjager) [inline peer/persistentpeer_mgr.go:62]: Maybe add a basic description of what is/makes a peer persistent?

File: peer/persistentpeer_mgr.go
Comment (joostjager) [inline peer/persistentpeer_mgr.go:421]: I think there may be a race condition with running goroutines created by the previous `ConnectPeer` that are blocked on the mutex?

File: peer/persistentpeer_mgr.go
Comment (joostjager) [inline peer/persistentpeer_mgr.go:600]: Isn't the `ConnectPeer` call doing its connects in a goroutine?

**Review by joostjager — COMMENTED** [review]
Did another pass of this PR focusing on concurrency issues first. I have to say that I find it hard to be certain that there are none with the extensive use of the lock, goroutines and cancel channel.

After my previous review, I mentioned:

>I wonder if it would be easier to understand with a single event loop per peer that receives updates via a channel.

I still think that would help a lot. Not just for review, but also for future devs working on this code.

Generally speaking, the connection handling for a peer is independent of all other peers. If that would be reflected in the code as a layer, it's already a bit easier to see how things work.

Then within the code for handling a single peer, I think that a single event loop that deals with retries, new addresses, backoff and additional connection requests will make the logic much more transparent.

File: peer/persistentpeer_mgr.go
Comment (Crypt-iQ) [reply peer/persistentpeer_mgr.go:421]: with `peer.connReqs`?

File: peer/persistentpeer_mgr.go
Comment (Crypt-iQ) [reply peer/persistentpeer_mgr.go:501]: Yes this seems possible, if the goroutine spawned by `ConnectPeer` doesn't get canceled by the chan and is waiting on the mutex lock

**Comment by ellemouton** [comment]
!lightninglabs-deploy mute

**Comment by ellemouton** [comment]
muting for a bit. Will pick this up again when my plate clears up a bit

**Comment by ellemouton** [comment]
closing for now. Can re-open once re-prio'd

**Pull request closed without merging**
